File: scripts/process_blockchain_jobs.ts
import "dotenv/config";
import { createClient as createSupabaseClient } from "@supabase/supabase-js";
import { createClient, createAccount } from "genlayer-js";
import { studionet } from "genlayer-js/chains";
import { TransactionStatus } from "genlayer-js/types";

type BlockchainJob = {
  id: string;
  job_type: "register_exam" | "commit_submission" | "finalize_score";
  exam_id: string | null;
  candidate_id: string | null;
  payload: Record<string, unknown>;
  status: string;
  retries: number;
};

const supabaseUrl = process.env.SUPABASE_URL!;
const supabaseServiceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY!;
const privateKey = process.env.GENLAYER_PRIVATE_KEY!;
const contractAddress = process.env.GENLAYER_CONTRACT_ADDRESS! as `0x${string}`;

if (!supabaseUrl || !supabaseServiceRoleKey || !contractAddress) {
  throw new Error("Missing required worker environment variables.");
}

if (!privateKey) {
  throw new Error("Missing GENLAYER_PRIVATE_KEY. Export it in the terminal before running the worker.");
}

const supabase = createSupabaseClient(supabaseUrl, supabaseServiceRoleKey);

const account = createAccount(privateKey as `0x${string}`);

const client = createClient({
  chain: studionet,
  account,
});

async function updateJob(
  jobId: string,
  values: Record<string, unknown>
) {
  const { error } = await supabase
    .from("blockchain_jobs")
    .update(values)
    .eq("id", jobId);

  if (error) {
    throw new Error(`Failed to update job ${jobId}: ${error.message}`);
  }
}

async function processRegisterExam(job: BlockchainJob) {
  const payload = job.payload as {
    examId: string;
    metadataHash: string;
    startTime: string;
    endTime: string;
  };

  const txHash = await client.writeContract({
    address: contractAddress,
    functionName: "create_exam",
    args: [
      payload.examId,
      payload.metadataHash,
      payload.startTime,
      payload.endTime,
    ],
    value: BigInt(0),
  });

  const receipt = await client.waitForTransactionReceipt({
    hash: txHash,
    status: TransactionStatus.ACCEPTED,
  });

  await updateJob(job.id, {
    status: "completed",
    tx_hash: txHash,
    contract_address: contractAddress,
    result: receipt,
    updated_at: new Date().toISOString(),
  });

  console.log(`Completed register_exam job: ${job.id}`);
}

async function processCommitSubmission(job: BlockchainJob) {
  const payload = job.payload as {
    examId: string;
    candidateId: string;
    submissionHash: string;
  };

  const txHash = await client.writeContract({
    address: contractAddress,
    functionName: "commit_submission",
    args: [
      payload.examId,
      payload.candidateId,
      payload.submissionHash,
    ],
    value: BigInt(0),
  });

  const receipt = await client.waitForTransactionReceipt({
    hash: txHash,
    status: TransactionStatus.ACCEPTED,
  });

  await updateJob(job.id, {
    status: "completed",
    tx_hash: txHash,
    contract_address: contractAddress,
    result: receipt,
    updated_at: new Date().toISOString(),
  });

  console.log(`Completed commit_submission job: ${job.id}`);
}

async function processFinalizeScore(job: BlockchainJob) {
  const payload = job.payload as {
    examId: string;
    candidateId: string;
    objectiveScore: number | string;
    subjectiveScore: number | string;
    totalScore: number | string;
  };

  const txHash = await client.writeContract({
    address: contractAddress,
    functionName: "finalize_score",
    args: [
      payload.examId,
      payload.candidateId,
      String(payload.objectiveScore),
      String(payload.subjectiveScore),
      String(payload.totalScore),
    ],
    value: BigInt(0),
  });

  const receipt = await client.waitForTransactionReceipt({
    hash: txHash,
    status: TransactionStatus.ACCEPTED,
  });

  await updateJob(job.id, {
    status: "completed",
    tx_hash: txHash,
    contract_address: contractAddress,
    result: receipt,
    updated_at: new Date().toISOString(),
  });

  console.log(`Completed finalize_score job: ${job.id}`);
}

async function processJob(job: BlockchainJob) {
  await updateJob(job.id, {
    status: "processing",
    updated_at: new Date().toISOString(),
  });

  try {
    if (job.job_type === "register_exam") {
      await processRegisterExam(job);
      return;
    }

    if (job.job_type === "commit_submission") {
      await processCommitSubmission(job);
      return;
    }

    if (job.job_type === "finalize_score") {
      await processFinalizeScore(job);
      return;
    }

    throw new Error(`Unknown job type: ${job.job_type}`);
  } catch (error) {
    const message =
      error instanceof Error ? error.message : "Unknown worker error";

    await updateJob(job.id, {
      status: "failed",
      error: message,
      retries: (job.retries || 0) + 1,
      updated_at: new Date().toISOString(),
    });

    console.error(`Failed job ${job.id}:`, message);
  }
}

async function main() {
  const { data, error } = await supabase
    .from("blockchain_jobs")
    .select("*")
    .eq("status", "pending")
    .order("created_at", { ascending: true })
    .limit(10);

  if (error) {
    throw new Error(`Failed to fetch pending jobs: ${error.message}`);
  }

  const jobs = (data || []) as BlockchainJob[];

  if (jobs.length === 0) {
    console.log("No pending blockchain jobs found.");
    return;
  }

  for (const job of jobs) {
    console.log(`Processing job ${job.id} (${job.job_type})`);
    await processJob(job);
  }
}

main().catch((error) => {
  console.error("Worker fatal error:", error);
  process.exit(1);
});

File: scripts/process_blockchain_jobs.py
import json
import subprocess
from datetime import datetime

from supabase import create_client

SUPABASE_URL = "https://xbpbsghhrcxxyzznffre.supabase.co"
SUPABASE_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]

CONTRACT_ADDRESS = "0xc08c7BcB3B10A587958473C97C26Ad654dF5fdDC"

supabase = create_client(SUPABASE_URL, SUPABASE_KEY)


def run_genlayer(args: list[str]) -> dict:
    result = subprocess.run(
        ["genlayer"] + args,
        cwd="/workspaces/examproof",
        capture_output=True,
        text=True,
    )
    return {
        "returncode": result.returncode,
        "stdout": result.stdout.strip(),
        "stderr": result.stderr.strip(),
    }


def process_job(job: dict):
    payload = job["payload"]
    job_type = job["job_type"]

    if job_type == "register_exam":
        args = [
            "write",
            CONTRACT_ADDRESS,
            "create_exam",
            json.dumps(payload["examId"]),
            json.dumps(payload["metadataHash"]),
            json.dumps(payload["startTime"]),
            json.dumps(payload["endTime"]),
        ]
    elif job_type == "commit_submission":
        args = [
            "write",
            CONTRACT_ADDRESS,
            "commit_submission",
            json.dumps(payload["examId"]),
            json.dumps(payload["candidateId"]),
            json.dumps(payload["submissionHash"]),
        ]
    elif job_type == "finalize_score":
        args = [
            "write",
            CONTRACT_ADDRESS,
            "finalize_score",
            json.dumps(payload["examId"]),
            json.dumps(payload["candidateId"]),
            json.dumps(str(payload["objectiveScore"])),
            json.dumps(str(payload["subjectiveScore"])),
            json.dumps(str(payload["totalScore"])),
        ]
    else:
        raise ValueError(f"Unknown job_type: {job_type}")

    result = run_genlayer(args)

    if result["returncode"] == 0:
        supabase.table("blockchain_jobs").update({
            "status": "completed",
            "result": result,
            "updated_at": datetime.utcnow().isoformat(),
            "contract_address": CONTRACT_ADDRESS,
        }).eq("id", job["id"]).execute()
    else:
        supabase.table("blockchain_jobs").update({
            "status": "failed",
            "error": result["stderr"] or result["stdout"],
            "result": result,
            "retries": (job.get("retries") or 0) + 1,
            "updated_at": datetime.utcnow().isoformat(),
        }).eq("id", job["id"]).execute()


def main():
    response = supabase.table("blockchain_jobs").select("*").eq("status", "pending").limit(20).execute()
    jobs = response.data or []

    for job in jobs:
        supabase.table("blockchain_jobs").update({
            "status": "processing",
            "updated_at": datetime.utcnow().isoformat(),
        }).eq("id", job["id"]).execute()

        try:
            process_job(job)
        except Exception as e:
            supabase.table("blockchain_jobs").update({
                "status": "failed",
                "error": str(e),
                "retries": (job.get("retries") or 0) + 1,
                "updated_at": datetime.utcnow().isoformat(),
            }).eq("id", job["id"]).execute()


if __name__ == "__main__":
    main()
